Add documentation for the "command" and "commandfor" attributes, and "command" event. (#30829)
mdn/content: files/en-us/web/api/invoker_commands_api/index.md

---
title: Invoker Commands API
slug: Web/API/Invoker_Commands_API
page-type: web-api-overview
status:
  - experimental
browser-compat:
  - api.CommandEvent
  - api.HTMLButtonElement.commandForElement
  - api.HTMLButtonElement.command
---

{{DefaultAPISidebar("Invoker Commands API")}}

> The **Invoker Commands API** provides a way to declaratively assign behaviors to buttons, allowing control of interactive elements when the button is enacted (clicked or invoked via a keypress, such as the spacebar or return key).

Reviewer: Perhaps "enacted" could be replaced with another word. It doesn't quite read naturally to me. Maybe "activated", idk what the normal word mdn uses is.

## Concepts and usage

A common pattern on the web is to have {{HTMLElement("button")}} elements control various aspects of the page, such as opening and closing {{domxref("Popover API", "popovers", "", "nocode")}} or {{HTMLElement("dialog")}} elements, formatting text, and more.

> Historically creating these kinds of controls has required JavaScript event listeners added to the button which can then call the APIs on the element they control. The {{domxref("HTMLButtonElement.commandForElement", "commandForElement")}} and {{domxref("HTMLButtonElement.command", "command")}} provide a way to do this declaratively for a limited set of actions. This can be advantageous for built-in commands as the user does not have to wait for JavaScript to download and execute to make these buttons interactive.

Reviewer: It should be The commandForElement and command **properties** provide a way...

Reviewer: Having said that it's not the properties that provide the way to do this declaratively it's the attributes so perhaps this should link to the attributes?

## HTML attributes

- [`commandfor`](/en-US/docs/Web/HTML/Element/button#commandfor)
  - : Turns a {{htmlelement("button")}} element into a control button, controlling the given interactive element; takes the ID of the element to control as its value.
- [`command`](/en-US/docs/Web/HTML/Element/button#command)
  - : Specifies the action to be performed on an element being controlled by a control `<button>`, specified via the `commandfor` attribute.

## Interfaces

- {{domxref("CommandEvent")}}
  - : Represents an event notifying the user that a command has been issued. It is the event object for the {{domxref("HTMLElement/command_event", "command")}} event. The event fires on element referenced by {{domxref("HTMLButtonElement.commandForElement", "commandForElement")}}.

## Extensions to other interfaces

### Instance properties

- {{domxref("HTMLButtonElement.commandForElement")}}

>   - : Gets and sets the element being controlled by the control button. The JavaScript equivalent of the [`commandfor`](/en-US/docs/Web/HTML/Element/button#commandfor) HTML attribute.

Reviewer: ```suggestion
  - : Gets and sets the element being controlled by the control button. The JavaScript equivalent of the [`commandfor`](/en-US/docs/Web/HTML/Element/button#commandfor) HTML attribute.
  + : Gets and sets the element being controlled by the button. The JavaScript equivalent of the [`commandfor`](/en-US/docs/Web/HTML/Element/button#commandfor) HTML attribute.
```

- {{domxref("HTMLButtonElement.command")}}
  - : Gets and sets the action to be performed on the element being controlled by the control button. Reflects the value of the [`command`](/en-US/docs/Web/HTML/Element/button#command) HTML attribute.

### Events

> - `HTMLElement` {{domxref("HTMLElement.command_event", "command")}} event

Reviewer: The `HTMLElement` bit feels superfluous idk if it's included in other pages?

  - : Fired on the element referenced by the control button.

## Examples

### Creating declarative popovers

```html

> <button commandfor="mypopover" command="toggle-popover">

Reviewer: We should be setting explicit type attributes in all these examples to encourage best practice (and avoid the oddities around forms)

---
title: Invoker Commands API
slug: Web/API/Invoker_Commands_API
page-type: web-api-overview
status:
  - experimental
browser-compat:
  - api.CommandEvent
  - api.HTMLButtonElement.commandForElement
  - api.HTMLButtonElement.command
---

{{DefaultAPISidebar("Invoker Commands API")}}

> The Invoker Commands API provides a way to declaratively assign behaviors to buttons, allowing control of interactive elements when the button is enacted (clicked or invoked via a keypress, such as the spacebar or return key).

Reviewer 2: ```suggestion
The **Invoker Commands API** provides a way to declaratively assign behaviors to buttons, allowing control of interactive elements when the button is enacted (clicked or invoked via a keypress, such as the spacebar or return key).
```

## Concepts and usage

> A common pattern on the web is to have {{domxref("HTMLButtonElement", "button")}} elements control various aspects of the page, such as opening and closing {{domxref("Popover API", "popovers", "", "nocode")}} or {{domxref("HTMLDialogElement", "dialog")}} elements, formatting text, and more.

Reviewer 2: ```suggestion
A common pattern on the web is to have {{HTMLElement("button")}} elements control various aspects of the page, such as opening and closing {{domxref("Popover API", "popovers", "", "nocode")}} or {{HTMLElement("dialog")}} elements, formatting text, and more.
```

General comment is we just link to HTML docs when we talk about the HTML elements (and we add the `<>`). Many HTML experts dislike it but it kind of sticks.

> Historically creating these kinds of controls has required JavaScript event listeners added to the button which can then call the APIs on the element they control. The {{domxref("HTMLButtonElement.commandForElement", "commandForElement")}} and {{domxref("HTMLButtonElement.command", "command")}} provide a way to do this declaratively for a limited set of actions.

Reviewer 3: @keithamus it might be nice to mention the benefit of this is as soon as the visitor can see the button, they can interact with it – rather than either having to wait, or the more common thing: seeing it rendered, thinking they can interact and wondering why it's not working (because the JavaScript powering it has yet to download/parse/execute).

> ## HTML Attributes

Reviewer 2: ```suggestion
## HTML attributes
```

> - [`commandfor`](/en-US/docs/Web/HTML/Element/button#commandfor):

Reviewer 2: ```suggestion
- [`commandfor`](/en-US/docs/Web/HTML/Element/button#commandfor)
```

  - : Turns a {{htmlelement("button")}} element into a control button, controlling the given interactive element; takes the ID of the element to control as its value.

> - {{domxref("HTMLButtonElement.command", "command")}} provide a way to do this declaratively for a limited set of actions.

Reviewer 2: If you want to point to the HTML content attributes, link to the HTML docs. You can use Markdown links because there are no macros for these.

> - [`command`](/en-US/docs/Web/HTML/Element/button#command):
> 
>   - : Specifies the action to be performed on an element being controlled by a control `<button>`, specified via the `commandfor` attribute. The possible values are:
> 
>     - `"show-modal"`
>       - : The button will show a {{htmlelement("dialog")}} as modal. If the dialog is already modal, no action will be taken.
>     - `"close"`
>       - : The button will close a {{htmlelement("dialog")}} element. If the dialog is already closed, no action will be taken.
>     - `"show-popover"`
>       - : The button will show a hidden popover. If you try to show an already showing popover, no action will be taken. See the {{domxref("Popover API", "Popover API", "", "nocode")}} landing page for more details.
>     - `"hide-popover"`
>       - : The button will hide a showing popover. If you try to hide an already hidden popover, no action will be taken. See the {{domxref("Popover API", "Popover API", "", "nocode")}} landing page for more details.
>     - `"toggle-popover"`
>       - : The button will toggle a popover between showing and hidden. If the popover is hidden, it will be shown; if the popover is showing, it will be hidden. See the {{domxref("Popover API", "Popover API", "", "nocode")}} landing page for more details.
>     - Custom values
>       - : This attribute can represent custom values that are prefixed with a two hyphen characters (`--`). Buttons with a custom value will dispatch the {{domxref("CommandEvent")}} on the controlled element.

Reviewer 2: ```suggestion
- [`command`](/en-US/docs/Web/HTML/Element/button#command)
  - : Specifies the action to be performed on an element being controlled by a control `<button>`, specified via the `commandfor` attribute.
```
